File: app/src/main/java/com/cn29/aac/ui/shopping/ArtistFragment.java
package com.cn29.aac.ui.shopping;


import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.cn29.aac.databinding.FragmentArtistBinding;
import com.cn29.aac.di.scope.AndroidDataBinding;
import com.cn29.aac.di.scope.ViewModel;
import com.cn29.aac.repo.itunes.Artist;
import com.cn29.aac.ui.base.BaseFragment;
import com.cn29.aac.ui.common.BaseRecyclerViewAdapter.OnItemClickListener;
import com.cn29.aac.ui.shopping.vm.ArtistFragmentViewModel;
import com.cn29.aac.ui.shopping.vm.ShoppingActivityViewModel;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import androidx.recyclerview.widget.GridLayoutManager;

public class ArtistFragment extends BaseFragment {

  @Inject
  @ViewModel
  ArtistFragmentViewModel viewModel;

  @Inject
  @ViewModel
  ShoppingActivityViewModel shoppingActivityViewModel;

  @Inject
  @AndroidDataBinding
  FragmentArtistBinding binding;

  private ArtistAdapter adapter;

  private List<Artist> list = new ArrayList<>();

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    shoppingActivityViewModel.getQueryData().observe(this, query -> {
          viewModel.getArtist(query).observe(this, response -> {
            switch (response.status) {
              case CHECKING:
                break;
              case SUCCESS:
                progressDialogComponent.hideLoading();
                list = response.data;
                OnItemClickListener<Artist> feedEntryOnItemClickListener = getFeedEntryOnItemClickListener();
                adapter = new ArtistAdapter(list,
                    feedEntryOnItemClickListener);
                binding.list.setAdapter(adapter);
                break;
              case ERROR:
                progressDialogComponent.hideLoading();
                break;
              case LOADING:
                progressDialogComponent.showLoading();
                break;
            }
          });
        }
    );
  }

  private OnItemClickListener<Artist> getFeedEntryOnItemClickListener() {
    return item ->
    {
      //open the artist detail
      Intent intent = new Intent(ArtistFragment.this.getActivity(), ArtistDetailActivity.class);
      intent.putExtra("artist", item);
      startActivity(intent);
    };
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    // Inflate the layout for this fragment
    binding.list.setLayoutManager(new GridLayoutManager(getActivity(), 4));
    return binding.getRoot();
  }

}

File: app/src/main/java/com/cn29/aac/ui/shopping/ArtistFragment.kt
package com.cn29.aac.ui.shopping

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.GridLayoutManager
import com.cn29.aac.databinding.FragmentArtistBinding
import com.cn29.aac.di.scope.AndroidDataBinding
import com.cn29.aac.di.scope.ViewModel
import com.cn29.aac.repo.itunes.Artist
import com.cn29.aac.repo.util.Resource
import com.cn29.aac.repo.util.Status
import com.cn29.aac.ui.base.BaseFragment
import com.cn29.aac.ui.common.BaseRecyclerViewAdapter
import com.cn29.aac.ui.shopping.vm.ArtistFragmentViewModel
import com.cn29.aac.ui.shopping.vm.ShoppingActivityViewModel
import java.util.*
import javax.inject.Inject

class ArtistFragment : BaseFragment() {
    @Inject
    @ViewModel
    lateinit var viewModel: ArtistFragmentViewModel

    @Inject
    @ViewModel
    lateinit var shoppingActivityViewModel: ShoppingActivityViewModel

    @Inject
    @AndroidDataBinding
    lateinit var binding: FragmentArtistBinding
    private var adapter: ArtistAdapter? = null
    private var list: List<Artist> = ArrayList()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        shoppingActivityViewModel
                .getQueryData()
                .observe(this,
                         Observer { query: String? ->
                             viewModel.getArtist(
                                             query)
                                     .observe(
                                             this,
                                             Observer { response: Resource<List<Artist>> ->
                                                 when (response.status) {
                                                     Status.CHECKING -> {
                                                     }
                                                     Status.SUCCESS -> {
                                                         progressDialogComponent!!.hideLoading()
                                                         list = response.data!!
                                                         val feedEntryOnItemClickListener = feedEntryOnItemClickListener
                                                         adapter = ArtistAdapter(
                                                                 list,
                                                                 feedEntryOnItemClickListener)
                                                         binding.list.adapter = adapter
                                                     }
                                                     Status.ERROR -> progressDialogComponent!!.hideLoading()
                                                     Status.LOADING -> progressDialogComponent!!.showLoading()
                                                 }
                                             })
                         }
                )
    }

    //open the artist detail
    private val feedEntryOnItemClickListener: BaseRecyclerViewAdapter.OnItemClickListener<Artist>
        get() = object : BaseRecyclerViewAdapter.OnItemClickListener<Artist> {
            override fun onItemClick(item: Artist) {
                val intent = Intent(
                        this@ArtistFragment.activity,
                        ArtistDetailActivity::class.java)
                intent.putExtra(
                        "artist",
                        item)
                startActivity(
                        intent)
            }

        }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        binding.list.layoutManager = GridLayoutManager(activity, 4)
        return binding.root
    }
}